File: src/Tests/StateUITests/ChannelTests.swift
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

// The path that describes nothing: a channel the platform moves many times a
// second, and the arithmetic that follows it.
//
// What has to hold, and each of these is one test below: a write records
// nothing and asks for no render; the value is still THERE for whoever reads
// it; a layout that follows one says so in its message, by a channel and a
// rule id; and the rule the id names answers the same placements a render
// would have described.
//
// The mechanism is in Core/Channel.swift; the host's half - which calls the
// rule on the platform's own frames and writes the numbers onto the controls -
// is StateUI.Runtime's Channels.cs.

import XCTest
@testable import StateUI

/// A view that reads a channel, so a test can see that reading one records
/// nothing.
private struct Follower: ContentView {
    let value: Channel<Double>
    let builds: Builds

    var content: Element {
        builds.count += 1
        return label("at \(value.wrappedValue)")
    }
}

/// Counts builds. A class, so the walk that collects state boxes leaves it
/// alone.
private final class Builds {
    var count = 0
}

final class ChannelTests: XCTestCase {
    override func setUp() {
        super.setUp()
        Renderer.shared.clearInvalidation()
        Renderer.shared.clearChannels()
    }

    // MARK: - The value

    /// Writing one asks for no render and names no change - which is the whole
    /// of what makes it affordable to move with a finger.
    func testWritingAChannelAsksForNoRender() {
        let value = Channel(wrappedValue: 0.0)

        value.wrappedValue = 40

        XCTAssertEqual(value.wrappedValue, 40)
        XCTAssertFalse(Renderer.shared.needsRender)
        XCTAssertTrue(Renderer.shared.pendingChanges.isEmpty)
    }

    /// And reading one records no dependency, so the view that read it is not
    /// rebuilt when it moves. The trade is the point: a view cannot SHOW one.
    func testReadingOneRecordsNothing() {
        let value = Channel(wrappedValue: 0.0)
        let builds = Builds()
        let renders = Renders()

        renders.render(stack([Follower(value: value, builds: builds).body], id: "root"))
        XCTAssertEqual(builds.count, 1)

        value.wrappedValue = 12
        renders.revisit(changed: Renderer.shared.pendingChanges)

        XCTAssertEqual(builds.count, 1)
    }

    /// The host says where it stands by the number it was issued, and the
    /// value is then what anything reading it sees - a handler asking where
    /// the run is, and the arithmetic itself.
    func testTheHostMovesItByItsChannel() {
        let value = Channel(wrappedValue: 0.0)
        let channel = value.channel

        Renderer.shared.moved(channel, to: 91.5)

        XCTAssertEqual(value.wrappedValue, 91.5)
    }

    /// A value is issued ONE number however often it is asked for it: the host
    /// quotes that number back, and a second one would be a second value.
    func testAChannelIsIssuedOnce() {
        let value = Channel(wrappedValue: 0.0)

        XCTAssertEqual(value.channel, value.channel)
        XCTAssertNotEqual(value.channel, Channel(wrappedValue: 0.0).channel)
    }

    // MARK: - What a message says about it

    /// A scroller told to report into a channel says so as a number, and no
    /// handler at all - there is nothing to run on this side.
    func testAScrollerNamesTheChannelItReportsInto() {
        let value = Channel(wrappedValue: 0.0)

        let node = ScrollView { Label("x") }
            .orientation(.horizontal)
            .scrollX(value)
            .body

        XCTAssertEqual(node.props[.scrollXChannel], .number(Double(value.channel)))
        XCTAssertNil(node.events[.scrollXChanged])
    }

    /// A view whose drag is written into values says both numbers.
    func testADraggedViewNamesTheValuesItIsWrittenInto() {
        let across = Channel(wrappedValue: 0.0)
        let down = Channel(wrappedValue: 0.0)

        let node = BoxView(Color("#000000")).panX(across).panY(down).body

        XCTAssertEqual(node.props[.panXChannel], .number(Double(across.channel)))
        XCTAssertEqual(node.props[.panYChannel], .number(Double(down.channel)))
    }

    /// A layout that follows channels carries their numbers AND an id for the
    /// arithmetic - the closure itself never leaves this side.
    func testAFollowingLayoutCarriesItsChannelsAndItsRule() {
        let across = Channel(wrappedValue: 0.0)
        let turn = Channel(wrappedValue: 0.0)
        let renders = Renders()

        func tree() -> Node {
            PlacedLayout([1, 2], id: \.self, following: across, turn, at: { index, _, _ in
                Placement(Rect(Double(index) * 10 + across.wrappedValue, 0, 20, 20))
            }) { number in
                Label("\(number)")
            }
            .id("run")
            .body
        }

        // The room is measured before anything can be placed in it, so the
        // layout itself is one frame late - exactly as it is without this.
        var patch = renders.render(tree())

        func reader(_ patch: Patch) -> Int? {
            if let id = patch.events?[.frameChanged] { return id }

            for child in patch.children where reader(child) != nil {
                return reader(child)
            }

            return nil
        }

        XCTAssertTrue(renders.fire(
            reader(patch) ?? -1, with: [.numbers([0, 0, 300, 100, 0, 0, 0, 0])]))

        patch = renders.renderFromScratch(tree())

        func layout(_ patch: Patch) -> Patch? {
            if patch.type == .absoluteLayout { return patch }

            for child in patch.children {
                if let found = layout(child) { return found }
            }

            return nil
        }

        let placed = try? XCTUnwrap(layout(patch))

        XCTAssertEqual(
            placed?.props[.channels],
            .numbers([Double(across.channel), Double(turn.channel)]))

        guard case .number(let rule)? = placed?.props[.channelRule] else {
            return XCTFail("the layout carries no rule id")
        }

        // AND THE ID NAMES THE ARITHMETIC, which is what the host calls on its
        // own frames. The same closure a render used: moving the value and
        // asking again answers the cards somewhere else, with nothing built.
        let arithmetic = try? XCTUnwrap(renders.placement(Int(rule)))

        XCTAssertEqual(arithmetic?(1, 2, Rect(0, 0, 300, 100)).bounds, Rect(10, 0, 20, 20))

        Renderer.shared.moved(across.channel, to: 25)

        XCTAssertEqual(arithmetic?(1, 2, Rect(0, 0, 300, 100)).bounds, Rect(35, 0, 20, 20))
    }

    /// The shade is a NUMBER to the host, and its absence is a number too: a
    /// layout with no shade view answers `unshaded`, which is the one value an
    /// opacity cannot be, and a layout with one answers what its arithmetic
    /// said. Without that the host could not tell a card wearing NONE of a
    /// shade from a run that has no shade at all, both of which say nought.
    func testAShadeSaysWhetherThereIsOneAtAll() {
        let across = Channel(wrappedValue: 0.0)

        func rule(_ shaded: Bool) -> PlacementRule? {
            let renders = Renders()

            var run = PlacedLayout([1, 2], id: \.self, following: across, at: { index, _, _ in
                Placement(Rect(0, 0, 20, 20), shade: Double(index) * 0.5)
            }) { number in
                Label("\(number)")
            }

            if shaded {
                run = run.shade(BoxView(Color("#000000")))
            }

            let patch = renders.render(run.id("run").body)

            func layout(_ patch: Patch) -> Patch? {
                if patch.type == .absoluteLayout { return patch }

                for child in patch.children {
                    if let found = layout(child) { return found }
                }

                return nil
            }

            // The room is measured before anything is placed in it, so the
            // first render carries the reader rather than the layout.
            func reader(_ patch: Patch) -> Int? {
                if let id = patch.events?[.frameChanged] { return id }

                for child in patch.children where reader(child) != nil {
                    return reader(child)
                }

                return nil
            }

            XCTAssertTrue(renders.fire(
                reader(patch) ?? -1, with: [.numbers([0, 0, 300, 100, 0, 0, 0, 0])]))

            let settled = renders.renderFromScratch(run.id("run").body)

            guard case .number(let id)? = layout(settled)?.props[.channelRule] else {
                XCTFail("the layout carries no rule id")
                return nil
            }

            return renders.placement(Int(id))
        }

        let room = Rect(0, 0, 300, 100)

        XCTAssertEqual(rule(false)?(1, 2, room).shade, PackedPlacement.unshaded, """
            a layout with no shade view says so in the one number an opacity \
            cannot be, whatever its arithmetic answered
            """)

        XCTAssertEqual(rule(true)?(0, 2, room).shade, 0, """
            a card wearing none of a shade the layout HAS says nought, which is \
            an opacity like any other
            """)

        XCTAssertEqual(rule(true)?(1, 2, room).shade, 0.5)
    }

    /// A shaded layout places a GRID whose second child is the shade, wearing
    /// the opacity the arithmetic answered - which is what the host writes onto
    /// between renders, and what the tree draws before it ever does.
    func testAShadedLayoutDrawsTheShadeOverEachPlacedView() {
        let renders = Renders()

        let run = PlacedLayout([1, 2], id: \.self, at: { index, _, _ in
            Placement(Rect(0, 0, 20, 20), opacity: 0.9, shade: Double(index) * 0.4)
        }) { number in
            Label("\(number)")
        }
        .shade(BoxView(Color("#000000")))

        let patch = renders.render(run.id("run").body)

        func layout(_ patch: Patch) -> Patch? {
            if patch.type == .absoluteLayout { return patch }

            for child in patch.children {
                if let found = layout(child) { return found }
            }

            return nil
        }

        func reader(_ patch: Patch) -> Int? {
            if let id = patch.events?[.frameChanged] { return id }

            for child in patch.children where reader(child) != nil {
                return reader(child)
            }

            return nil
        }

        XCTAssertTrue(renders.fire(
            reader(patch) ?? -1, with: [.numbers([0, 0, 300, 100, 0, 0, 0, 0])]))

        let settled = renders.renderFromScratch(run.id("run").body)

        guard let placed = layout(settled),
              let first = placed.children.first,
              let near = first.children.last,
              let last = placed.children.last,
              let far = last.children.last
        else {
            return XCTFail("the run placed no views")
        }

        XCTAssertEqual(first.type, .grid, """
            a shaded layout places a grid of two - the view and the shade over \
            it - because the placement goes on one node and the shade's own \
            opacity on another
            """)
        XCTAssertEqual(first.children.count, 2)
        XCTAssertEqual(first.props[.opacity], .number(0.9), "the placement's own fade")
        XCTAssertEqual(near.props[.opacity], .number(0))
        XCTAssertEqual(far.props[.opacity], .number(0.4), "the far view wears more of the shade")
    }

    /// A layout given no value to follow says neither, so nothing on the far
    /// side has anything to call.
    func testALayoutFollowingNothingSaysNothing() {
        let node = following(AbsoluteLayout { Label("x") }, [], nil)

        XCTAssertNil(node.props[.channels])
        XCTAssertNil(node.placing)
    }

    // MARK: - The reader

    /// A `ScrollReader` lays an empty scroller over what it holds, as long as
    /// the room plus how far the run goes beyond it, reporting into the
    /// channel.
    func testAScrollReaderReportsIntoItsChannel() {
        let across = Channel(wrappedValue: 0.0)
        let renders = Renders()

        let patch = renders.render(
            ScrollReader(across: 540) { Label("under") }
                .scrollX(across)
                .snapInterval(90)
                .id("reader")
                .body)

        func scroller(_ patch: Patch) -> Patch? {
            if patch.type == .scrollView { return patch }

            for child in patch.children {
                if let found = scroller(child) { return found }
            }

            return nil
        }

        let found = scroller(patch)

        XCTAssertEqual(found?.props[.scrollXChannel], .number(Double(across.channel)))
        XCTAssertEqual(found?.props[.snapInterval], .number(90))
        XCTAssertEqual(found?.props[.orientation]?.enumeration, ScrollOrientation.horizontal.rawValue)
    }
}

File: src/Tests/StateUIRuntime.Tests/ChannelTests.cs
// SPDX-FileCopyrightText: 2026 Paweł Krzywdziński and Contributors
// SPDX-License-Identifier: Apache-2.0

// The host's half of a CHANNEL: a value the platform moves many times a second,
// the layouts that follow it, and the ground a follower's write stands on.
//
// The arithmetic itself is Swift's, so what can be asked here is everything
// AROUND it - where a value stands, what a described layout owes, and what this
// side is allowed to remember. A host with no Swift module registered answers
// no placements at all, which is exactly the state these run in.

using StateUI.Runtime.Rendering;

namespace StateUI.Runtime.Tests;

public class ChannelTests
{
    /// <summary>
    /// Where a value stands is kept whether or not anything follows it: a drag
    /// MOVES a value rather than setting it, so where it began has to be known,
    /// and a render that happens later describes the views where the reader
    /// left them.
    /// </summary>
    [Fact]
    public void AValueThatMovedIsWhereItWasLastSaidToBe()
    {
        var channels = new Channels(new MotionEngine());

        Assert.Equal(0, channels.Standing(7));

        channels.Moved(7, 12.5);
        Assert.Equal(12.5, channels.Standing(7));

        channels.Moved(7, -3);
        Assert.Equal(-3, channels.Standing(7));

        Assert.Equal(0, channels.Standing(8));
    }

    /// <summary>
    /// A DESCRIBED LAYOUT'S DELTA IS SPENT. The tree has just said where
    /// everything is, and what it said already holds wherever the reader has
    /// moved the run to - the arithmetic reads the channels. So each placed
    /// view goes back to the translation the TREE wrote, never to the one this
    /// side wrote on top of it.
    /// </summary>
    [Fact]
    public void ADescribedLayoutGoesBackToTheTranslationTheTreeWrote()
    {
        var channels = new Channels(new MotionEngine());
        var layout = new AbsoluteLayout();
        var card = new BoxView();

        layout.Children.Add(card);
        channels.Follows(layout, [4.0], rule: 9);

        // What the tree said, as it said it.
        channels.Authored(card, x: 12);
        channels.Authored(card, y: -5);

        // And where this side then moved it, following the reader's hand.
        card.TranslationX = 240;
        card.TranslationY = 60;

        channels.Applied(layout);

        Assert.Equal(12, card.TranslationX);
        Assert.Equal(-5, card.TranslationY);
    }

    /// <summary>
    /// A view the tree never gave a translation goes back to nothing, which is
    /// MAUI's own answer for a property nobody set.
    /// </summary>
    [Fact]
    public void AViewTheTreeNeverMovedGoesBackToNothing()
    {
        var channels = new Channels(new MotionEngine());
        var layout = new AbsoluteLayout();
        var card = new BoxView { TranslationX = 88, TranslationY = 4 };

        layout.Children.Add(card);
        channels.Follows(layout, [4.0], rule: 9);
        channels.Applied(layout);

        Assert.Equal(0, card.TranslationX);
        Assert.Equal(0, card.TranslationY);
    }

    /// <summary>
    /// A layout nobody registered is left alone: an apply reaches every layout
    /// on the page, and only a FOLLOWED one owes an alignment.
    /// </summary>
    [Fact]
    public void ALayoutThatFollowsNothingIsLeftWhereItIs()
    {
        var channels = new Channels(new MotionEngine());
        var layout = new AbsoluteLayout();
        var card = new BoxView { TranslationX = 88 };

        layout.Children.Add(card);
        channels.Applied(layout);

        Assert.Equal(88, card.TranslationX);
    }

    /// <summary>
    /// A FOLLOWED LAYOUT SAYS SO, which is what the arranger's measure reads:
    /// such a layout is the size it is GIVEN, its children standing where
    /// arithmetic over the room puts them.
    /// </summary>
    [Fact]
    public void AFollowedLayoutIsMarkedAsOne()
    {
        var channels = new Channels(new MotionEngine());
        var layout = new AbsoluteLayout();

        Assert.False(layout.GetValue(Channels.FollowedProperty) is true);

        channels.Follows(layout, [1.0, 2.0], rule: 5);

        Assert.True(layout.GetValue(Channels.FollowedProperty) is true);
    }

    /// <summary>
    /// A render RE-REGISTERS what it already had, and the rule id is the one
    /// thing that can move - so a layout is followed once per value however
    /// many times it is described.
    /// </summary>
    [Fact]
    public void ALayoutDescribedAgainIsFollowedOnce()
    {
        var channels = new Channels(new MotionEngine());
        var layout = new AbsoluteLayout();
        var card = new BoxView();

        layout.Children.Add(card);

        channels.Follows(layout, [1.0], rule: 5);
        channels.Follows(layout, [1.0], rule: 5);
        channels.Follows(layout, [1.0], rule: 6);

        // The alignment an apply owes still finds exactly one registration -
        // it would throw or double-write if the list had grown.
        channels.Authored(card, x: 3);
        channels.Applied(layout);

        Assert.Equal(3, card.TranslationX);
    }

    /// <summary>
    /// NOTHING HERE OUTLIVES A PAGE. A session lasts for the process and every
    /// page is left eventually, so a layout that once followed a value - and
    /// every view the tree ever gave a translation - must be collectable the
    /// moment the tree stops holding it.
    /// </summary>
    [Fact]
    public void AFollowerAndItsViewsAreLetGoOfWhenThePageIs()
    {
        var channels = new Channels(new MotionEngine());

        (WeakReference Layout, WeakReference Card) held = Registered(channels);

        GC.Collect();
        GC.WaitForPendingFinalizers();
        GC.Collect();

        Assert.False(held.Layout.IsAlive, "the layout is the page's, not the session's");
        Assert.False(held.Card.IsAlive, "and so is every view it placed");
    }

    /// <summary>Registers a layout and lets go of it, keeping only a watch.</summary>
    private static (WeakReference Layout, WeakReference Card) Registered(Channels channels)
    {
        var layout = new AbsoluteLayout();
        var card = new BoxView();

        layout.Children.Add(card);

        channels.Follows(layout, [11.0], rule: 2);
        channels.Authored(card, x: 7);
        channels.Applied(layout);

        return (new WeakReference(layout), new WeakReference(card));
    }
}
